in_toto/keylib.go

```diff
@@ -2,220 +2,402 @@ package in_toto
+	if err != nil {
+		return err
 	}
 
-	if keyObj.KeyVal.Private == "" {
-		return keyObj, fmt.Errorf("this key is not a private key")
+	// pam.Decode returns the parsed pem block and a rest.
+	// The rest is everything, that could not be parsed as PEM block.
+	// Therefore we can drop this via using the blank identifier "_"
+	data, _ := pem.Decode(pemBytes)
+	if data == nil {
+		return ErrNoPEMBlock
 	}
 
-	// 64 hexadecimal digits => 32 bytes for the private portion of the key
-	if len(keyObj.KeyVal.Private) != 64 {
-		return keyObj, fmt.Errorf("the private field on this key is malformed")
+	// Try to load private key, if this fails try to load
+	// key as public key
+	key, err := ParseKey(data.Bytes)
+	if err != nil {
+		return err
 	}
 
-	return keyObj, nil
+	// Use type switch to identify the key format
+	switch key.(type) {
+	case *rsa.PublicKey:
+		if err := k.SetKeyComponents(pemBytes, []byte{}, "rsa", scheme, keyIdHashAlgorithms); err != nil {
+			return err
+		}
+	case *rsa.PrivateKey:
+		// Note: We store the public key as PKCS8 key here, although the private key get's stored as PKCS1 key
+		// This behavior is consistent to the securesystemslib
+		pubKeyBytes, err := x509.MarshalPKIXPublicKey(key.(*rsa.PrivateKey).Public())
+		if err != nil {
+			return err
+		}
+		if err := k.SetKeyComponents(pubKeyBytes, pemBytes, "rsa", scheme, keyIdHashAlgorithms); err != nil {
+			return err
+		}
+	case ed25519.PublicKey:
+		if err := k.SetKeyComponents(key.(ed25519.PublicKey), []byte{}, "ed25519", scheme, keyIdHashAlgorithms); err != nil {
+			return err
+		}
+	case ed25519.PrivateKey:
+		pubKeyBytes := key.(ed25519.PrivateKey).Public()
+		if err := k.SetKeyComponents(pubKeyBytes.(ed25519.PublicKey), key.(ed25519.PrivateKey), "ed25519", scheme, keyIdHashAlgorithms); err != nil {
+			return err
+		}
+	case *ecdsa.PrivateKey:
+		pubKeyBytes, err := x509.MarshalPKIXPublicKey(key.(*ecdsa.PrivateKey).Public())
+		if err != nil {
+			return err
+		}
+		if err := k.SetKeyComponents(pubKeyBytes, pemBytes, "ecdsa", scheme, keyIdHashAlgorithms); err != nil {
+			return err
+		}
+	case *ecdsa.PublicKey:
+		if err := k.SetKeyComponents(pemBytes, []byte{}, "ecdsa", scheme, keyIdHashAlgorithms); err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("%w: %T", ErrUnsupportedKeyType, key)
+	}
+	return nil
 }
 
 /*
-GenerateEd25519Signature creates an ed25519 signature using the key and the
-signable buffer provided. It returns an error if the underlying signing library
-fails.
-*/
-func GenerateEd25519Signature(signable []byte, key Key) (Signature, error) {
+GenerateSignature will automatically detect the key type and sign the signable data
+with the provided key. If everything goes right GenerateSignature will return
+a for the key valid signature and err=nil. If something goes wrong it will
+return an not initialized signature and an error. Possible errors are:
 
-	var signature Signature
+	* ErrNoPEMBlock
+	* ErrUnsupportedKeyType
```

Comment:
Even though we only support one default signature scheme, we should check if the scheme on the key matches it.

Reply:
done in the latest commits